src/fluree/db/api/query.cljc

```diff
@@ -75,38 +77,64 @@
-
-            asserts-chan   (json-ld-resp/flakes->res db cache compact fuel 1000000
-                                                     {:wildcard? true, :depth 0}
-                                                     0 assert-flakes)
-            retracts-chan  (json-ld-resp/flakes->res db cache compact fuel 1000000
-                                                     {:wildcard? true, :depth 0}
-                                                     0 retract-flakes)
-
-            asserts (<? asserts-chan)
-            retracts (<? retracts-chan)
-
-            ;; t is always positive for users
-            result         (cond-> {:t (- (flake/t (first t-flakes)))}
-                             asserts (assoc :assert asserts)
-                             retracts (assoc :retract retracts))]
-        result)
+      (let [json-chan (json-ld-resp/flakes->res db cache compact fuel 1000000
+                                                {:wildcard? true, :depth 0}
+                                                0 s-flakes)]
+        (-> (<? json-chan)
+            ;; add the id in case the iri flake isn't present in s-flakes
+            (assoc :id (json-ld/compact (<? (dbproto/-iri db (flake/s (first s-flakes)))) compact))))
       (catch* e
               (log/error e "Error converting history flakes.")
               (async/>! error-ch e)))))
 
+(defn t-flakes->json-ld
+  [db compact cache fuel error-ch t-flakes]
+  (go-try
+    (let [assert-flakes (not-empty (filter flake/op t-flakes))
+          retract-flakes (not-empty (filter (complement flake/op) t-flakes))
+
+          s-asserts-ch (->> (sort-by flake/s assert-flakes)
+                            (partition-by flake/s)
+                            (async/to-chan!))
+          s-retracts-ch (->> (sort-by flake/s retract-flakes)
+                             (partition-by flake/s)
+                             (async/to-chan!))
+
+          s-asserts-out-ch (async/chan)
+          s-retracts-out-ch (async/chan)
+
+          s-asserts-json-ch (async/into [] s-asserts-out-ch)
+          s-retracts-json-ch (async/into [] s-retracts-out-ch)]
+      ;; process asserts
+      (async/pipeline-async 2
+                            s-asserts-out-ch
+                            (fn [assert-flakes ch]
+                              (-> (s-flakes->json-ld db cache compact fuel error-ch assert-flakes)
+                                  (async/pipe ch)))
+                            s-asserts-ch)
+      ;; process retracts
+      (async/pipeline-async 2
+                            s-retracts-out-ch
+                            (fn [retract-flakes ch]
+                              (-> (s-flakes->json-ld db cache compact fuel error-ch retract-flakes)
+                                  (async/pipe ch)))
+                            s-retracts-ch)
+      {(json-ld/compact const/iri-t compact) (- (flake/t (first t-flakes)))
+       (json-ld/compact const/iri-assert compact) (<? s-asserts-json-ch)
```

Comment:
`<?` can throw errors, but `s-asserts-json-ch` will either fail to close or close without an error. you should use `<!` instead. The same goes for the line below.